### ISSUE TITLE
upstream xclim.core.units?

### ISSUE THREAD

**Issue by dcherian** [Ouranosinc/xclim#780]
Hello,

It looks like large parts of [`xclim.core.units`](https://xclim.readthedocs.io/en/stable/_modules/xclim/core/units.html0) could be upstreamed to cf-xarray and pint-xarray. Would you be interested in doing this?

`pin2cfunits` looks particularly useful as a custom format for "dequantifying" pint-backed xarray objects (cc @keewis).
https://github.com/Ouranosinc/xclim/blob/6102e542e6e08072a60879d6200f9340207cd50e/xclim/core/units.py#L167

**Comment by aulemahal** [comment]
Yes! That is one of our long-term goals. Currently, all unit handling is done "by hand" inside xclim, which is suboptimal. I think moving a much of what we can upstream is a great idea.

The registry modifications here are also based on MetPy, but augmented as other issues arose.

I don't think we will be able to remove all unit handling from xclim, but switching to integrated units will be a plus.  I'll try to look into this when I have time, but I don't think anyone from the core team will have much time to work on this in the next month.

**Comment by keewis** [comment]
posted first as https://github.com/jbusecke/cmip6_preprocessing/issues/167#issuecomment-939440765: FYI the custom unit formats PR was merged into `pint` so now it should be possible to register `xclim.core.units.pint2cfunits` as `"cfunits"` or `"cf"` and get something like `f"{quantity:.3f#cfunits}"` to work (documentation is still missing, though)

see hgrecco/pint#1375 and hgrecco/pint#1371

**Comment by dcherian** [comment]
Nice work @keewis . This looks like a good time to upstream this to `cf_xarray.units`!

**Comment by keewis** [comment]
just tried it:
```python
In [7]: import pint
   ...: 
   ...: pint.formatting._FORMATTERS.pop("cf", None)
   ...: 
   ...: @pint.register_unit_format("cf")
   ...: def pint2cfunits(unit, registry, **options) -> str:
   ...:     """Return a CF-compliant unit string from a `pint` unit.
   ...: 
   ...:     Parameters
   ...:     ----------
   ...:     unit : pint.UnitContainer
   ...:         Input unit.
   ...:     registry : pint.UnitRegistry
   ...:         the associated registry
   ...:     **options
   ...:         Additional options (may be ignored)
   ...: 
   ...:     Returns
   ...:     -------
   ...:     out : str
   ...:         Units following CF-Convention, using symbols.
   ...:     """
   ...:     import re
   ...: 
   ...:     # convert UnitContainer back to Unit
   ...:     unit = registry.Unit(unit)
   ...:     # Print units using abbreviations (millimeter -> mm)
   ...:     s = f"{unit:~D}"
   ...: 
   ...:     # Search and replace patterns
   ...:     pat = r"(?P<inverse>/ )?(?P<unit>\w+)(?: \*\* (?P<pow>\d))?"
   ...: 
   ...:     def repl(m):
   ...:         i, u, p = m.groups()
   ...:         p = p or (1 if i else "")
   ...:         neg = "-" if i else ("^" if p else "")
   ...: 
   ...:         return f"{u}{neg}{p}"
   ...: 
   ...:     out, n = re.subn(pat, repl, s)
   ...: 
   ...:     # Remove multiplications
   ...:     out = out.replace(" * ", " ")
   ...:     # Delta degrees:
   ...:     out = out.replace("Δ°", "delta_deg")
   ...:     return out.replace("percent", "%")
   ...: 
   ...: ureg = pint.application_registry
   ...: u = ureg.Unit("m ** 2 / (s ** 3 * kg)")
   ...: display(u)
   ...: display(f"{u:cf}")
<Unit('meter ** 2 / kilogram / second ** 3')>
'm^2 kg-1 s-3'
```
which I think should do the same as `pint2cfunits`.

The caret seems a bit odd to me (I was assuming it would return `m2 kg-1 s-3`) but this seems intentional and I don't know that much about cfunits.

**Comment by aulemahal** [comment]
Indeed that's nice!

So as for upstreaming, I understand that for the moment it concerns the following elements:
- Unit and dimensionalities definitions.
    * Does this include the "hydro" context? This is something we use to convert water flux units without having to explicitly specify the water's density. But it might be too magic for cf-xarray.
-  `pint2cfunits` registered has shown above.
- `units2pint`  : to read the (cf) units of a DataArray into a pint object.

Other tools of `xclim.core.units` are for unit _manipulation_ and AFAIU this is left to pint[-xarray] for the moment?

**Comment by tlogan2000** [comment]
closing for #1010
